<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
  <style type="text/css">
    tr.d0 td {
    background-color: #CCCCFF; color: black;
    }
    tr.d1 td {
    background-color: #CCFFCC; color: black;
    }
  </style>


  <A name="intro"><H1>Introduction</H1></A>
  <P>
    RIVALS is a software tool that allows for fast range queries and fast-in order traversals of genomic interval graphs.  Many common tasks that are performed on interval graphs can be performed efficiently when the intervals are sorted in order by start position.  Examples of these tasks include pooling different interval graphs into one, finding areas of maximal coverage, and locating islands of connected reads.  RIVALS contains "runners", which apply a certain function (such as merge, overlap, flatten) to a sorted interval graph and produce another sorted interval graph as the result.  By chaining these runners together, RIVALS can perform complex analyses on very large genomic datasets.
  </P>
  <P>
    Genomic data is stored in an augmented interval tree.  The data is stored on disk as a contiguous list of intervals sorted by start position, thus making efficient use of cache.  Each interval consists of a chromosome (or scaffold/contig), strand (+, -, both), start position (inclusive), and stop position (exclusive).  Searching the interval graph for a range of intervals can be done efficiently by viewing the interval list as a binary tree.  Each interval in the tree is augmented with the longest endpoint of its subtree.  By adding this extra field to the intervals, analysis can be done on either the entire dataset or over a particular range of interest.
  </P>
  <P>
   [TODO: Write something about Lua?]
  </P>
  
  <A name="install"><H1>Installation</H1></A>
  <A name="quick"><H1>Quick Start</H1></A>
  <p>
    Example scripts are included with the software to quickly get started using its features.  
  </p>

  <A name="usage"><H1>Usage</H1></A>
  <A name="import"><H1>Import</H1></A>

  <A name="format"><H1>File Format</H1></A>
  <P>
    When a file is imported into a RIVAL sample, two files are created: &lt;base name&gt;.riv and &lt;base name&gt;.map, where "base name" is the name given to the sample during import.  The .riv file is a binary file containing the list of imported intervals sorted by start position.  The .map file is a human-readable text file listing the chromosome names and indices.  The .map file contains three tab-delimited fields: chromosome name, position in list of chromosome start (inclusive), position in list of chromosome stop (exclusive).  These files should not be edited, with the sole exception of changing the names of the chromosomes, which is explained next. 
  </P>
  <UL>
    <A name="alias"><H2>Changing Chromosome Names</H2></A>
  <P>
    A chromosome can have various names depending on the naming convention (e.g. "1", "chr1", "NC_000001" for chromosome 1 of human).  When performing RIVAL tasks that combines multiple samples (such as merge or overlap), it is essential that the chromosome names agree.  RIVALS will use the names given in the .map file.  You may change the names in these files so that the chromosome naming convention is consistent across all samples of interest, such as below.

    <blockquote>
      <em>Changing chromosome names in ncbi.map</em><br \>
      <table border="1">
	<tr>
	  <td>
	    <code>
	      NC_000001    244625    398054<br \>
	      NC_000010   1223225   1304380<br \>
	      NC_000011   1042658   1133814<br \>
	      NC_000012   398054    472108<br \>
	      .......
	    </code>
	  </td>
	  <td>
	    <H4>
	      Change to<br \>
	      <center>=></center>
	    </H4>
	  </td>
	  <td>
	    <code>
	      chr1    244625    398054<br \>
	      chr10   1223225   1304380<br \>
	      chr11   1042658   1133814<br \>
	      chr12   398054    472108<br \>
	      .......
	    </code>
	  </td>
	</tr>
      </table>
    </blockquote>
    Now the sample known as "ncbi" will match any sample that contains the naming convention chr1, chr2, etc.
  </P>
  </UL>
  
  <A name="rivals"><H1>RIVAL Commands</H1></A>
  <P>
    This section includes a comprehensive list of commands available in RIVALS.  The commands can roughly be divided into categories as follows: importers, sources, runners, and sinks.  Importers are functions that take a text file and converts it into a RIVALS file.  Currently, BED and GFF files are supported.  The source commands allow you to load interval graphs.  You can either create an interval manually or load an interval graph that you imported from a BED or GFF file.  Once loaded, you chain together the runners to perform analyses on the interval graph.  Finally, the sinks complete the analysis by outputting the results or reporting a summary of the results.  The following table gives an overview of the different commands.  Clicking on a command will give you a more details on what the command does.  Please visit the Examples section to see examples of the commands in action. 
  </P>

  <table border="2">
    <tr>
      <th colspan="3">Importers: commands that convert text files into RIVAL files</th>
    </tr>
    <tr class="d0">
      <td><code>string <A HREF="manual.html#install" target="manual">importBED</A></code></td>
      <td>("from_bed", "to_rival" [,overwrite=false])</td>
      <td>Import BED file into RIVAL file<br />Only overwrites if optional argument is true</td>
    </tr>
    <tr class="d1">
      <td><code>string <A HREF="manual.html#install" target="manual">importGFF</A></code></td>
      <td>("from_gff", "to_rival" [,overwrite=false])</td>
      <td>Import GFF file into RIVAL file<br />Only overwrites if optional argument is true</td>
    </tr>
    <tr>
      <th colspan="3">Sources: commands that create an initial Runner</th>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">interval</A></code></td><td>("chr", start, stop)</td>
      <td>Create the interval chr:start-stop</td>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">range</A></code></td><td>("from_rival")</td>
      <td>Return all intervals in RIVAL file</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">range</A></code></td><td>("from_rival", "chr")</td>
      <td>Return all intervals in RIVAL file that are on chromosome chr</td>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">range</A></code></td><td>("from_rival", "chr", start, stop)</td>
      <td>Return all intervals in RIVAL file that overlap chr:start-stop</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">range</A></code></td><td>("from_rival", "chr", point)</td>
      <td>Return all intervals in RIVAL file that overlap chr:point</td>
    </tr>
    <tr>
      <th colspan="3">Runners: commands that can be linked together</th>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">merge</A></code></td><td>(Runner A, Runner B)</td>
      <td>Merge intervals from A and B</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">flatten</A></code></td><td>(Runner A)</td>
      <td>Join overlapping intervals from A</td>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">clique</A></code></td><td>(Runner A, int minimum_overlap)</td>
      <td>Report regions with at least minimum_overlap depth</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">overlaps</A></code></td><td>(Runner A, Runner B)</td>
      <td>Report intervals in A that overlap some interval in B</td>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">contained_in</A></code></td><td>(Runner A, Runner B)</td>
      <td>Report intervals in A that are contained in some interval in B</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">contains</A></code></td><td>(Runner A, Runner B)</td>
      <td>Report intervals in A that contain some interval in B</td>
    </tr>
    <tr class="d1">
      <td><code>Runner <A HREF="manual.html#install" target="manual">get_strand</A></code></td><td>(Runner A, "strand")</td>
      <td>Report intervals on A that are from the specified strand: "+", "-", or "." (both)</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">set_strand</A></code></td><td>(Runner A, "strand")</td>
      <td>Set intervals on A to the specified strand: "+", "-", or "." (both)</td>
    </tr>
    <tr class="d0">
      <td><code>Runner <A HREF="manual.html#install" target="manual">score_cutoff</A></code></td><td>(Runner A, int threshold)</td>
      <td>Reports all intervals with score greater than or equal to the threshold</td>
    </tr>
    <tr class="d1">
      <th colspan="3">Sinks: commands that summarize or output the list of intervals</th>
    </tr>
    <tr class="d0">
      <td><code>string <A HREF="manual.html#install" target="manual">saveAsBED</A></code></td><td>(Runner A, "to_bed")</td>
      <td>Create a BED file from runner A</td>
    </tr>
    <tr class="d1">
      <td><code>string <A HREF="manual.html#install" target="manual">saveAsBED</A></code></td><td>(Runner A)</td>
      <td>Print runner A in BED format to standard out</td>
    </tr>
    <tr class="d0">
      <td><code>string <A HREF="manual.html#install" target="manual">saveAsRival</A></code></td><td>(Runner A, "to_rival")</td>
      <td>Create a RIVAL file from runner A</td>
    </tr>
    <tr class="d1">
      <td><code>number <A HREF="manual.html#install" target="manual">countIntervals</A></code></td><td>(Runner A)</td>
      <td>Count the number of intervals in runner A</td>
    </tr>
  </table>

  <H2>Runners</H2>
  <UL>
    <A name="alias"><H2>merge</H2></A>
    <P>
      <img src="Figures/merge.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>flatten</H2></A>
    <P>
      <img src="Figures/flatten.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>clique</H2></A>
    <P>
      <img src="Figures/clique.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>overlaps</H2></A>
    <P>
      <img src="Figures/overlaps.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>contained_in</H2></A>
    <P>
      <img src="Figures/contained_in.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>contains</H2></A>
    <P>
      <img src="Figures/contains.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>get_strand</H2></A>
    <P>
      <img src="Figures/get_strand.png" alt="FigFlatten" />
    </P>
  </UL>
  <UL>
    <A name="alias"><H2>set_strand</H2></A>
    <P>
      <img src="Figures/set_strand.png" alt="FigFlatten" />
    </P>
  </UL>

  <A name="examples"><H1>Examples</H1></A>
  <H2>Importing and Exporting data</H2>
  <P>
    Assume the files "sample1.bed" and "sample2.gff" exist in your current directory.  Import them as "bed_sample" and "gff_sample" using the BED and GFF importers, respectively:
    <blockquote>
      <code>
	-- Import bed and gff files into Rival samples<br />
	importBED("sample1.bed", "bed_sample")<br />
	importGFF("sample2.gff", "gff_sample")<br />
	<br />
      </code>
    </blockquote>
	
    Save the Rival sample "gff_sample" to a BED file.
    <blockquote>
      <code>
	-- Save only chromosome 8 of gff_sample to a BED file<br />
	saveAsBED("gff_to_bed.bed", range("gff_sample", "chr8"))<br />
      </code>
    </blockquote>
  </P>
  
  <H2>Finding Enriched Islands</H2>
  <P>
    An island is a set of alignments that span a block of contiguous bases.  If you view the genomic intervals as a graph - where each node is an alignment and edges are formed between overlapping alignments - then an island is a set of nodes where each node is connected to each other node by some path in the graph.  Below are some examples of finding islands with high read coverage.
  </P>
      
  <P>
    One measure of an island is the "population" of the island - the number of alignments that belong to the island.  You can find this by using the flatten command, followed by a score cutoff (20 in this example).
    <blockquote>
      <code>
	-- Find all islands containing at least 20 total alignments<br />
	sample = importBED("nrsf.bed", "nrsf")<br />
	saveAsBED(score_cutoff(flatten(range(sample)), 20))
      </code>
    </blockquote>
  </P>
  <P>
    Another measure of an island is the "elevation" of the island - the maximum coverage of any of the bases in the island.
    <blockquote>
      <code>
	-- Find all islands containing a position of at least 100x coverage<br />
	saveAsBED( contains( flatten(range(sample)), clique(range(sample), 100)))
      </code>
    </blockquote>
  </P>
  
  <H2>Retrieving Regions of Interest</H2>
  <P>
    This example shows how to count the number of reads that overlap a particular region of interest.  It shows two valid approaches, with one more efficient than the other.<br /><br />
    
    <em>Method 1: Linear Pass</em>
    <blockquote>
      <code>
	A = range("stat1")<br />
	B = interval("chr1", 13863058, 13864058)<br />
	print(count(overlaps(A, B)))<br />
      </code><br />
    </blockquote>

    <em>Method 2: Augmented Interval Tree</em><br />
    <blockquote>
      <code>
	A = range("stat1", "chr1", 13863058, 13864058)<br />
	print(count(A))<br />
      </code>
    </blockquote>
    Both give the same result.  The first method does a linear pass over the intervals on chromosome 1 and counts the number of intervals that overlap the region of interest.  Method two uses an interval tree to quickly locate the intervals that overlap the region.  The interval tree will outperform the linear pass when the number of intervals in the sample is large and when the number of overlaps is much less than the total intervals. 
  </P>
  
  <H1>Data Storage</H1>
  <P>
    <img src="Figures/data_storage.png"  width="400" alt="DataStorage" />
    <br />
    Each interval contains a start position, stop position, strand, and maximum endpoint of subtree (used for range queries).  Strand is 2 bits and is placed within the start and stop fields.  The start, stop and max_end fields are (b-1) bits each (with 1 unused bit in the max_end field).  Each interval uses 3*b bits (by default, b=32 and the fields are signed integers).
  </P>

  <H1>Data Traversal</H1>
  <P>
    <img src="Figures/data_traversal.png"  width="500" alt="DataTraversal" />
    <br />
    A.  The largest endpoint of the subtree (12) is smaller than the start point (19) of the query interval and so does not have to be searched.  B.  The start point of the node interval (35) is larger than the endpoint (30) of the query interval so the subtree of the right child does not need to be searched.  C.  Intervals are arranged in contiguous blocks on disk by start position which allows for quickly and efficient in-order traversals. 
  </P>
  
  <A name="authors"><H1>Authors</H1></A>
  <p>
    RIVALS is written by Justin Brown and Lina Dong.  For questions or feedback, please email run.intervals@gmail.com
  </p>
  <A name="license"><H1>License</H1></A>
  <p>
    Copyright 2010 Justin T. Brown<br />
    All Rights Reserved<br />
    EMAIL: run.intervals@gmail.com
  </p>

  <p>
    Rivals is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
  </p>
  
  <p>
    Rivals is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.
  </p>

</HTML>
